File: 0125-valid-palindrome/0125-valid-palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        Deque<Character> strs = new ArrayDeque<>();
        
        for (char c: s.toCharArray()) {
            if (Character.isAlphabetic(c) || Character.isDigit(c)) {
                strs.add(Character.toLowerCase(c));
            }
        }
        
        while (strs.size() > 1) {
            if (strs.pollFirst() != strs.pollLast()) {
                return false;
            }
        }
        
        return true;
    }
}

File: 0125-valid-palindrome/0125-valid-palindrome.kt
class Solution {
    fun isPalindrome(s: String): Boolean {
        var start = 0
        var end = s.length - 1
        
        while (start < end) {
            when {
                !Character.isLetterOrDigit(s[start]) -> start++
                !Character.isLetterOrDigit(s[end]) -> end--
                else -> {
                    if (Character.toLowerCase(s[start]) != Character.toLowerCase(s[end])) {
                        return false
                    }
                    start++
                    end--
                }
            }
        }
        
        return true
    }
}